storybookjs/storybook
# Vite: Re-export essentials sub-addon preview annotations for pnpm

Issue: #19176

Closes https://github.com/storybookjs/storybook/pull/19530

## What I did

In order to support pnpm from the vite builder, I added re-exports of the sub-addons which have preview annotations, and adjusted the preset handling to leave bare imports alone instead of turning them into absolute paths.  This allows pnpm to correctly find the files and avoids users from having to "shamefully hoist" or add the sub-addons to their package.json. 

I added `typesVersions` to the addons, so that we can `export * from "@storybook/addon-actions/preview"`, for instance.  The .d.ts file isn't at the root of the package, it's in `/dist`, so using typesVersions lets us point to the right spot.  This is one of the approaches shown in https://github.com/andrewbranch/example-subpath-exports-ts-compat, and has the best overall compatibility.

One problem I hit is that tsup generates empty .d.ts files for `manager`, so I filed https://github.com/egoist/tsup/issues/762 and needed to ignore errors when re-exporting from them.

## How to test

We don't have any good tests for pnpm, unfortunately, so this is a bit manual.

- Build this branch, especially `lib/builder-vite`, `lib/core-common`, and `addons/essentials`
- Somewhere else on your filesystem, create a pnpm vite project (`pnpm create vite@latest --template=react`)
- Add storybook (`pnpx sb@next init --builder=vite`)
- Copy over the `dist` of `lib/builder-vite` into `node_modules/.pnpm/@storybook+builder-vite@7.0.0-alpha.XX_<hash>/node_modules/@storybook/builder-vite`
- Copy over the `dist` of `lib/core-common` into `node_modules/.pnpm/@storybook+core-common@7.0.0-alpha.XX_<hash>/node_modules/@storybook/core-common`
- Copy the `dist` and `package.json` of `addons/essentials` into `node_modules/@storybook/addon-essentials/`
- Start up storybook, (`pnpm run storybook`), and it should start successfully.
